File: [0235][Easy][Lowest Common Ancestor of a Binary Search Tree]/Lowest_Common_Ancestor_of_a_Binary_Search_Tree_2.java
//Definition for a binary tree node.
class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode(int x) { val = x; }
} 
// solution 2, use iteration to travel all the nodes of the BST.
// leetcode time     cost : 9 ms
// leetcode memory   cost : 39.4 MB
// Time  Complexity: O(N)
// Space Complexity: O(1)
class Solution {
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {

        // Value of p
        int pVal = p.val;

        // Value of q;
        int qVal = q.val;

        // Start from the root node of the tree
        TreeNode node = root;

        // Traverse the tree
        while (node != null) {

            // Value of ancestor/parent node.
            int parentVal = node.val;

            if (pVal > parentVal && qVal > parentVal) {
                // If both p and q are greater than parent
                node = node.right;
            } else if (pVal < parentVal && qVal < parentVal) {
                // If both p and q are lesser than parent
                node = node.left;
            } else {
                // We have found the split point, i.e. the LCA node.
                return node;
            }
        }
        return null;
    }
}

File: [0235][Easy][Lowest Common Ancestor of a Binary Search Tree]/Lowest_Common_Ancestor_of_a_Binary_Search_Tree_2.py
#-*- coding: utf-8 -*-  
# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None
# solution 2, use iteration to travel all the nodes of the BST.
# leetcode time     cost : 80 ms
# leetcode memory   cost : 20 MB
# Time  Complexity: O(N)
# Space Complexity: O(1)
class Solution:
    def lowestCommonAncestor(self, root, p, q):
        """
        :type root: TreeNode
        :type p: TreeNode
        :type q: TreeNode
        :rtype: TreeNode
        """

        # Value of p
        p_val = p.val

        # Value of q
        q_val = q.val

        # Start from the root node of the tree
        node = root

        # Traverse the tree
        while node:

            # Value of current node or parent node.
            parent_val = node.val

            if p_val > parent_val and q_val > parent_val:    
                # If both p and q are greater than parent
                node = node.right
            elif p_val < parent_val and q_val < parent_val:
                # If both p and q are lesser than parent
                node = node.left
            else:
                # We have found the split point, i.e. the LCA node.
                return node